Sources/NIOCore/Docs.docc/ByteBuffer-lengthPrefix.md

```diff
@@ -0,0 +1,119 @@
+# Writing length-prefixed data in ByteBuffer
+
+This article explains how to write data prefixed with a length, where the length could be encoded in various ways.
+
+## Overview
+
+We often need to write some data prefixed by its length. Sometimes, this may simply be a fixed width integer. But many
+protocols encode the length differently, depending on how big it is. For example, the QUIC protocol uses variable-length
+integer encodings, in which smaller numbers can be encoded in fewer bytes. The first 2 bits of the first byte indicate
+how many further bytes should be read.
+
+We have added functions to help with reading and writing data which is prefixed with lengths encoded by various
+strategies.
+
+## ``NIOBinaryIntegerEncodingStrategy`` protocol
+
+The first building block is a protocol which describes how to encode and decode an integer.
+
+An implementation of this protocol is needed for any encoding strategy. Currently, we provide one
+implementation: ``ByteBuffer/QUICBinaryEncodingStrategy``
+
+This protocol only has 2 requirements which don't have default implementations:
+
+- `readInteger`: Reads an integer from the ByteBuffer using this encoding. Implementations will read as many bytes as
+  they need to, according to their wire format, and move the reader index accordingly
+- `writeInteger`: Write an integer to the ByteBuffer using this encoding. Implementations will write as many bytes as
+  they need to, according to their wire format, and move the writer index accordingly.
+
+Note that implementations of this protocol need to either:
+
+- Encode the length of the integer into the integer itself when writing, so it knows how many bytes to read when
+  reading. This is what QUIC does.
+- Always use the same length, e.g. a simple strategy which always writes the integer as a UInt64.
+
+## Extensions on ``ByteBuffer``
+
+To provide a more user-friendly API, we have added extensions on `ByteBuffer` for writing integers with a
+chosen ``NIOBinaryIntegerEncodingStrategy``. These are ``ByteBuffer/writeEncodedInteger(_:strategy:)``
+and ``ByteBuffer/readEncodedInteger(_:as:)``.
```

Comment:
Feels a bit odd to have:

```swift
buffer.writeEncodedInteger(42, strategy: .quic)
```

vs.

```swift
buffer.readEncodedInteger(.quic: as Int32.self)
```

IMO the latter would make more sense as:

```swift
// Explicit as
buffer.readEncodedInteger(as: Int32.self, strategy: .quic)
// Or, inferring the type:
buffer.readEncodedInteger(strategy: .quic)
```